WEBVTT
1
00:00:00.300 --> 00:00:01.960
Welcome to Workbook 8.2.

2
00:00:01.980 --> 00:00:07.260
I already went ahead and launch the starter project and the first thing it tells us is that in the previous

3
00:00:07.260 --> 00:00:12.270
workbook we were handling bad requests by throwing and catching checked exceptions.

4
00:00:12.300 --> 00:00:20.280
This was really annoying because we had to go and modify every single method to throw that checked exception,

5
00:00:20.280 --> 00:00:23.160
modify the service interface.

6
00:00:23.160 --> 00:00:29.490
And then we had to try to run the code in every single handler method here and catch the exception if

7
00:00:29.490 --> 00:00:30.500
the code fails.

8
00:00:30.510 --> 00:00:36.510
We had to modify the code in many places of our app for a single functionality, and that's not something

9
00:00:36.510 --> 00:00:40.130
that I like in this lesson or in this workbook, I should say.

10
00:00:40.140 --> 00:00:44.760
We're going to handle bad requests inside of a single controller advice class.

11
00:00:45.060 --> 00:00:50.070
Let's start with Task One tells us to create a new folder called Exception, where we're going to create

12
00:00:50.070 --> 00:00:54.370
a contact not found exception, class and paste in the following code.

13
00:00:54.390 --> 00:00:59.220
Let's go ahead and do that right now and we'll create a new folder called Exception.

14
00:01:00.850 --> 00:01:09.100
I'll create a new file called No Contact Found Exception dot java and I will just paste in the code

15
00:01:09.100 --> 00:01:10.420
from learn the part.

16
00:01:11.350 --> 00:01:12.610
And did I?

17
00:01:12.850 --> 00:01:13.290
Oh.

18
00:01:15.620 --> 00:01:16.880
Should rename this.

19
00:01:17.900 --> 00:01:18.470
Okay.

20
00:01:18.470 --> 00:01:23.840
So the reason this is a runtime exception is so that we're not forced to catch it wherever it's being

21
00:01:23.840 --> 00:01:24.400
thrown.

22
00:01:24.410 --> 00:01:30.110
Whenever you throw a new instance of an exception, it's constructor is going to be called.

23
00:01:30.110 --> 00:01:36.170
We're expecting whoever throws a new instance of this exception to pass in an ID because it's also going

24
00:01:36.170 --> 00:01:39.690
to pass in a custom message into our parent constructor.

25
00:01:39.710 --> 00:01:41.480
Now, what does this do behind the scenes?

26
00:01:41.480 --> 00:01:48.050
If you're curious, if we go to this parent constructor, which passes the message into another parent

27
00:01:48.050 --> 00:01:51.780
constructor, which passes the message into another parent constructor.

28
00:01:51.800 --> 00:01:55.610
What this does is it updates a field called detail message.

29
00:01:55.610 --> 00:02:00.680
So the detail message field is going to contain the message that we ultimately passed in.

30
00:02:00.680 --> 00:02:05.750
Our message can be accessed using the gap message function, which is exactly what we're going to do.

31
00:02:05.750 --> 00:02:09.470
But anyways, no need to focus too much on the nitty gritty.

32
00:02:09.470 --> 00:02:15.440
All I want you to know is that whenever this exception is thrown in this manner, throw new contact,

33
00:02:15.470 --> 00:02:16.730
not found exception.

34
00:02:16.730 --> 00:02:22.760
When it's caught inside of our controller advice, we'll be able to get this message as follows.

35
00:02:22.760 --> 00:02:24.920
Exception get message.

36
00:02:26.050 --> 00:02:27.070
All right.

37
00:02:28.060 --> 00:02:33.820
Going back to task two, we need to modify the defined index by ID function to throw this newly created

38
00:02:33.820 --> 00:02:34.720
exception.

39
00:02:34.960 --> 00:02:43.000
Now, if we go to the service implementation, our find index by the function already throws an unchecked

40
00:02:43.000 --> 00:02:46.600
exception in the name of no such element exception.

41
00:02:46.600 --> 00:02:54.970
But in this case I just wanted to throw a custom exception instead a custom exception that is more descriptive

42
00:02:54.970 --> 00:03:01.420
and recall the constructor for this exception expects an ID as this exception is being created.

43
00:03:01.420 --> 00:03:07.730
It's going to pass a custom message into the parent constructor, allowing us to access it later on.

44
00:03:07.750 --> 00:03:12.580
And finally, because this is a runtime exception, we're not going to be forced to try and catch it.

45
00:03:12.580 --> 00:03:15.630
So we don't need to specify that this function throws anything.

46
00:03:15.640 --> 00:03:20.290
This makes our lives much easier because now we do not have to touch any of these operations.

47
00:03:20.290 --> 00:03:22.720
We do not have to touch the interface.

48
00:03:22.720 --> 00:03:26.170
We do not have to do anything inside of the controller.

49
00:03:26.170 --> 00:03:31.750
Any exception that gets thrown during the runtime, we're going to handle inside of a controller advice

50
00:03:31.750 --> 00:03:32.530
class.

51
00:03:32.530 --> 00:03:34.300
We'll get to that momentarily.

52
00:03:34.330 --> 00:03:38.830
Task three is to run the application and make a get request from postmen.

53
00:03:38.830 --> 00:03:40.030
Let's do just that.

54
00:03:40.030 --> 00:03:42.610
I'll copy this over to Postman right now.

55
00:03:42.640 --> 00:03:45.190
It seemed I'd already done this earlier, but whatever.

56
00:03:45.190 --> 00:03:46.480
I'll just press send again.

57
00:03:47.740 --> 00:03:50.080
And we get an internal server error.

58
00:03:50.350 --> 00:03:53.600
Now here the consumer is sending a bad request.

59
00:03:53.620 --> 00:03:56.170
The ID they're sending over does not exist.

60
00:03:56.170 --> 00:04:02.980
And as a result, the application fails and it throws an exception, an unchecked exception during the

61
00:04:02.980 --> 00:04:10.060
runtime that throws our runtime exception, and then it responds with a 500 status code.

62
00:04:10.390 --> 00:04:17.380
The 500 status code must be reserved for actual failures due to unexpected conditions.

63
00:04:17.380 --> 00:04:23.830
Like, for example, if our application is not able to connect to a database for some unexpected reason,

64
00:04:23.830 --> 00:04:25.380
the service might be down.

65
00:04:25.390 --> 00:04:32.710
It will send back a status of 500, but the consumer sending a bad ID is something that should be totally

66
00:04:32.710 --> 00:04:33.490
expected.

67
00:04:33.490 --> 00:04:40.660
US as developers, we need to anticipate this scenario and ensure that the application does not fail.

68
00:04:40.690 --> 00:04:48.430
It must be fully equipped to handle the bad request and respond with a 400 status code, which implies

69
00:04:48.430 --> 00:04:49.540
that it's their fault.

70
00:04:49.600 --> 00:04:52.240
They are the ones who passed in an invalid ID.

71
00:04:52.270 --> 00:04:59.950
So we'll go over here and it tells us that the controller advice is a class level annotation that allows

72
00:04:59.950 --> 00:05:02.980
you to define global exception handlers.

73
00:05:03.040 --> 00:05:09.100
I'll move on to Task four, which tells us to create a class called Application Exception Handler and

74
00:05:09.100 --> 00:05:11.590
apply the annotation we just talked about.

75
00:05:12.010 --> 00:05:12.940
All right.

76
00:05:13.980 --> 00:05:15.900
To let me collapse all of these.

77
00:05:17.010 --> 00:05:22.620
And I'll create a class called Application Exception Handler.

78
00:05:22.920 --> 00:05:28.450
This will serve as the global exception handler across our entire application.

79
00:05:28.470 --> 00:05:32.070
Anything that gets thrown, we're going to handle right here.

80
00:05:32.100 --> 00:05:32.610
All right.

81
00:05:32.610 --> 00:05:38.280
And inside of this controller advice, we can define exception handlers.

82
00:05:38.400 --> 00:05:44.730
The exception handler is a method level annotation that instruments a method to handle exceptions.

83
00:05:44.730 --> 00:05:49.860
So inside of our global exception handler, we're going to create an exception handler method called

84
00:05:49.860 --> 00:05:52.140
handle contact, not found exception.

85
00:05:52.920 --> 00:05:54.510
So I'm going to say public.

86
00:05:55.610 --> 00:06:03.530
And when a contact is not found, when this exception gets thrown, what we want to do is send back

87
00:06:03.530 --> 00:06:05.360
a response entity.

88
00:06:07.010 --> 00:06:09.200
That encapsulates a status code.

89
00:06:10.950 --> 00:06:12.140
HTTP status.

90
00:06:12.150 --> 00:06:17.810
As you may have guessed it, that's going to end up being 404 and we're going to call it handle contact,

91
00:06:17.820 --> 00:06:19.170
not found exception.

92
00:06:19.710 --> 00:06:24.450
And this method is going to receive the contact, not found exception that gets thrown.

93
00:06:25.170 --> 00:06:32.580
Now, in order for this method to actually get invoked, when this exception gets thrown, we need to

94
00:06:32.580 --> 00:06:40.830
market as an exception handler that is capable of handling exceptions of type contact, not found exception

95
00:06:41.460 --> 00:06:42.180
class.

96
00:06:43.410 --> 00:06:44.280
All right.

97
00:06:45.410 --> 00:06:51.350
And so when this runtime exception gets thrown, instead of our app failing during the runtime, it's

98
00:06:51.350 --> 00:06:54.230
actually going to get handled inside of this method.

99
00:06:54.230 --> 00:07:03.440
And we're just going to return a new response entity with a status of 404, which would be not found.

100
00:07:05.460 --> 00:07:07.740
All right, let's test this out.

101
00:07:08.740 --> 00:07:10.450
And this is a lot better.

102
00:07:11.660 --> 00:07:12.120
Well, right.

103
00:07:12.120 --> 00:07:16.140
And here it tells us it would be nice to send the user an error response.

104
00:07:16.140 --> 00:07:19.110
So right now it just sends back a 404.

105
00:07:19.110 --> 00:07:21.720
But let's send back some JSON.

106
00:07:21.720 --> 00:07:25.500
What we're going to do is right over here, define.

107
00:07:26.870 --> 00:07:29.540
An error response pod class.

108
00:07:29.540 --> 00:07:35.150
So the idea is that we'll be creating an object out of this class and all of its properties are going

109
00:07:35.150 --> 00:07:37.190
to be serialized into JSON.

110
00:07:37.190 --> 00:07:46.940
So here I'll just say private string message and I will generate a complete constructor as well as getters

111
00:07:46.940 --> 00:07:47.990
and setters.

112
00:07:54.530 --> 00:07:55.820
Here, I'll create.

113
00:07:56.930 --> 00:07:59.300
A new object.

114
00:08:00.510 --> 00:08:02.460
Of the error response class.

115
00:08:04.860 --> 00:08:09.480
And the message we're going to pass in is exception, but get message.

116
00:08:10.280 --> 00:08:11.960
So when we threw.

117
00:08:13.150 --> 00:08:16.390
Our contact not found exception and passed in an IED.

118
00:08:17.000 --> 00:08:23.860
Our exception passed in a message into the parent constructor which updated a particular field behind

119
00:08:23.860 --> 00:08:24.730
the scenes.

120
00:08:25.920 --> 00:08:30.750
And the exceptional get message method is able to grab the contents of that field.

121
00:08:31.200 --> 00:08:31.740
All right.

122
00:08:31.740 --> 00:08:39.030
And now here it tells us to actually specify a return type of object, because object is the most generic

123
00:08:39.030 --> 00:08:40.289
Java class.

124
00:08:41.200 --> 00:08:43.159
We can put anything over here.

125
00:08:43.179 --> 00:08:46.810
What we're going to do is send back an error response.

126
00:08:48.370 --> 00:08:50.500
As well as in HTTP status code.

127
00:08:51.010 --> 00:08:58.270
Now this error response, this object spring boot is going to serialize all of its properties into JSON

128
00:08:58.270 --> 00:09:00.580
before sending it back to the consumer.

129
00:09:01.060 --> 00:09:02.710
So if we try this out.

130
00:09:07.140 --> 00:09:14.790
Press send our our response was serialized into JSON such that the message property equals a JSON value

131
00:09:14.820 --> 00:09:17.770
of the ID 1 to 3 does not exist in our records.

132
00:09:17.790 --> 00:09:19.560
Very, very nice.

133
00:09:20.330 --> 00:09:21.230
All right.

134
00:09:21.560 --> 00:09:23.740
And I think we already did that.

135
00:09:23.750 --> 00:09:27.710
We have a 404 not found task eight.

136
00:09:27.740 --> 00:09:30.020
It would be nice to also send back a time stamp.

137
00:09:30.020 --> 00:09:31.880
Let's do just that.

138
00:09:35.540 --> 00:09:38.330
Local date time meets import it.

139
00:09:39.330 --> 00:09:45.810
And as soon as the error gets created, we're just going to set the time stamp equal to the current

140
00:09:45.810 --> 00:09:46.590
time.

141
00:09:46.590 --> 00:09:49.560
So I think we can do that by setting local date time.

142
00:09:49.890 --> 00:09:50.610
Now.

143
00:09:52.370 --> 00:09:53.110
Looks good.

144
00:09:53.120 --> 00:09:57.620
Let's add the usual getters and setters to our pod class.

145
00:10:00.470 --> 00:10:01.250
Okay.

146
00:10:03.080 --> 00:10:09.590
And now as we send back the error response, Spring Boot will serialize these two properties into JSON.

147
00:10:12.050 --> 00:10:13.370
It does just that.

148
00:10:13.370 --> 00:10:14.910
But I don't know about you.

149
00:10:14.930 --> 00:10:17.990
This last part is not very readable.

150
00:10:17.990 --> 00:10:26.660
So going over to Task nine, it tells us that as the property is being serialized into JSON, we can

151
00:10:26.660 --> 00:10:31.190
use the JSON format annotation to configure how it's serialized.

152
00:10:32.520 --> 00:10:39.210
So here, as you can see, our time stamp is already being serialized into a JSON string, so we're

153
00:10:39.210 --> 00:10:40.740
going to do just that.

154
00:10:41.840 --> 00:10:47.360
But the pattern we're going to serialize this date to is going to be the following pattern.

155
00:10:47.360 --> 00:10:52.430
So I'm just going to go ahead and copy this into our code.

156
00:10:57.690 --> 00:10:58.950
Trying this again.

157
00:11:00.450 --> 00:11:02.460
And this is just beautiful.

158
00:11:02.490 --> 00:11:09.570
Our API is now resilient to bad requests and it tells the user exactly what they did wrong.

159
00:11:09.660 --> 00:11:14.370
As a consumer, this tells me exactly what I need to know to make a good request.

